staging/src/k8s.io/apiserver/pkg/util/version/version.go

```diff
@@ -0,0 +1,223 @@
+	components := strings.Split(s, ".")
+	if len(components) != 2 {
+		return fmt.Errorf("version %s is not in the format of major.minor", s)
+	}
+	ver, err := version.ParseGeneric(s)
+	if err != nil {
+		return err
+	}
+	v.Val.Store(ver)
+	return nil
+}
+
+// String returns the flag value
+func (v *VersionVar) String() string {
+	ver := v.Val.Load()
+	return ver.String()
+}
+
+// Type gets the flag type
+func (v *VersionVar) Type() string {
+	return "version"
+}
+
+type effectiveVersion struct {
+	binaryVersion atomic.Pointer[version.Version]
+	// If the emulationVersion is set by the users, it could only contain major and minor versions.
+	// In tests, emulationVersion could be the same as the binary version, or set directly,
+	// which can have "alpha" as pre-release to continue serving expired apis while we clean up the test.
+	emulationVersion VersionVar
+	// minCompatibilityVersion could only contain major and minor versions.
+	minCompatibilityVersion VersionVar
+}
+
+func (m *effectiveVersion) BinaryVersion() *version.Version {
+	return m.binaryVersion.Load()
+}
+
+func (m *effectiveVersion) EmulationVersion() *version.Version {
+	// Emulation version can have "alpha" as pre-release to continue serving expired apis while we clean up the test.
+	// The pre-release should not be accessible to the users.
+	return m.emulationVersion.Val.Load().WithPreRelease(m.BinaryVersion().PreRelease())
+}
+
+func (m *effectiveVersion) MinCompatibilityVersion() *version.Version {
+	return m.minCompatibilityVersion.Val.Load()
+}
+
+func (m *effectiveVersion) Set(binaryVersion, emulationVersion, minCompatibilityVersion *version.Version) {
+	m.binaryVersion.Store(binaryVersion)
+	m.emulationVersion.Val.Store(emulationVersion)
+	m.minCompatibilityVersion.Val.Store(minCompatibilityVersion)
+}
+
+func (m *effectiveVersion) SetBinaryVersionForTests(binaryVersion *version.Version, featureGate featuregate.FeatureGate) func() {
+	oldBinaryVersion := m.binaryVersion.Load()
+	m.Set(binaryVersion, binaryVersion, binaryVersion.SubtractMinor(1))
+	oldFeatureGateVersion := featureGate.(featuregate.MutableVersionedFeatureGate).EmulationVersion()
+	if err := featureGate.(featuregate.MutableVersionedFeatureGate).SetEmulationVersion(binaryVersion); err != nil {
+		panic(err)
+	}
+	return func() {
+		m.Set(oldBinaryVersion, oldBinaryVersion, oldBinaryVersion.SubtractMinor(1))
+		featureGate.(featuregate.MutableVersionedFeatureGateForTests).Reset()
+		if err := featureGate.(featuregate.MutableVersionedFeatureGate).SetEmulationVersion(oldFeatureGateVersion); err != nil {
+			panic(err)
+		}
+	}
+}
+
+func (m *effectiveVersion) Validate(featureGate featuregate.FeatureGate) []error {
+	var errs []error
+	// Validate only checks the major and minor versions.
+	binaryVersion := m.binaryVersion.Load().WithPatch(0)
+	emulationVersion := m.emulationVersion.Val.Load().WithPatch(0)
+	minCompatibilityVersion := m.minCompatibilityVersion.Val.Load().WithPatch(0)
```

Comment:
Are there use cases where we expect or need patch to be non-zero?  I'm curious if we should clear non-zero values on `Set` or so we have a guarantee in all the code that consumes these versions that there won't be a patch version to consider.

I suspect we never want patch version on emulation and min compatibility version.  I'm less clear on if binary version should retain patch version..

Reply:
Moved `WithPatch(0)` to `Set`. I think we should keep the binary version as it is to keep the fidelity of what it means.